keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,858 @@
+|---------------|------------------|------------------------------------------------------------------------------------------------|
+| `host`        | string           | Host name to connect to.                                                                       |
+| `httpHeaders` | HTTPHeader array | Custom headers to set in the request.                                                          |
+| `path`        | string           | Path to access on the HTTP server.                                                             |
+| `port`        | string / int     | Name or number of the port to access on the container. Number must be in the range 1 to 65535. |
+| `scheme`      | string           | Scheme to use for connecting to the host: `HTTP`, `HTTPS`. (default: `"HTTP"`)                 |
+
+The `httpHeaders` field has type [`HTTPHeader array`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpheader-v1-core)
+with the following specification:
+
+(NOTE: We will need to evaluate how feasible it is to extend `HTTPHeader v1 core` with `valueFrom` to enable mounting headers from Secrets,
+for authentication. This may need to be its own independent KEP.)
+
+| Field       | Type             | Description                                                            |
+|-------------|------------------|------------------------------------------------------------------------|
+| `name`      | string           | The header field name                                                  |
+| `value`     | string           | The header field value                                                 |
+| `valueFrom` | HTTPHeaderSource | Source for the header's value. Cannot be used if `value` is non-empty. |
+
+The `valueFrom` field has type `HTTPHeaderSource` with the following specification:
+
+| Field          | Type              | Description                                             |
+|----------------|-------------------|---------------------------------------------------------|
+| `secretKeyRef` | SecretKeySelector | Selects a key of a secret in the ReplicaSet's namespace |
+
+The `secretKeyRef` field has type [`SecretKeySelector`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#secretkeyselector-v1-core):
+
+| Field  | Type   | Description                                                       |
+|--------|--------|-------------------------------------------------------------------|
+| `key`  | string | The key of the secret to select from. Must be a valid secret key. |
+| `name` | string | Name of the referent.                                             |
+| `key`  | string | Specify whether the Secret or its key must be defined.            |
+
+
+#### "Pod Picker" API contract:
+
+The contract for a `downscalePodPicker` API will be as follows:
+
+- Request payloads to the API will contain:
+   - `number_of_pods_requested` (`int`): minimum number of Pods to return
```

Comment:
why does this buck all k8s trends?  Shouldn't it at least follow most of the k8s API conventions?

Reply:
Can you please be more clear, about what API conventions this is not adhering to?

Reply:
allNamingConventions for one.  Also self-describing with metadata, including a version.  Assuming we go down this road, which I am not ready to assume, why would this not look like most other k8s objects?

Reply:
These are not "k8s objects" so I am confused, they are simply the payloads that exchange data between a Pod-Picker API and the kube-controller (and are implemented as REST or gRPC APIs).

I am not against using camel-case if that's your concern, that would simply make the fields:
- REQUEST payloads sent by kube-controller:
   - `numberOfPodsRequested`
   - `candidatePods`
- RESPONSE payloads sent by the Pod-Picker API:
   - `chosenPods`
   - `tiedPods`

Reply:
@alculquicondor it's easy to use camel case if that's what people think is best.

Regarding versioning, I agree in general, but in this case, I think we can avoid introducing an explicit version until we actually make our first "significant change" (with the implied default being v1 if a version field is ever added).

Especially as in this case, I don't think we are likely to ever make a truly "breaking change" given the simplicity of the API contract.